cpp/include/raft/core/resource/device_memory_resource.hpp

```diff
@@ -41,36 +66,161 @@ class device_memory_resource : public resource {
  */
 class workspace_resource_factory : public resource_factory {
  public:
-  workspace_resource_factory(rmm::mr::device_memory_resource* mr_ = nullptr) : mr(mr_) {}
-  resource_type get_resource_type() override { return resource_type::WORKSPACE_RESOURCE; }
-  resource* make_resource() override { return new device_memory_resource(mr); }
+  explicit workspace_resource_factory(
+    std::shared_ptr<rmm::mr::device_memory_resource> mr = {nullptr},
+    std::optional<std::size_t> allocation_limit         = std::nullopt,
+    std::optional<std::size_t> alignment                = std::nullopt)
+    : allocation_limit_(allocation_limit.value_or(default_allocation_limit())),
+      alignment_(alignment),
+      mr_(mr ? mr : default_plain_resource())
+  {
+  }
+
+  auto get_resource_type() -> resource_type override { return resource_type::WORKSPACE_RESOURCE; }
+  auto make_resource() -> resource* override
+  {
+    return new limiting_memory_resource(mr_, allocation_limit_, alignment_);
+  }
+
+  /** Construct a sensible default pool memory resource. */
+  static inline auto default_pool_resource(std::size_t limit)
```

Comment:
Instead of encoding certain assumptions here, I would just remove `default_pool_resource` together with `set_workspace_to_pool_resource`. I think it is sufficient to have `set_workspace_resource` API, the user could set a pool using that. An example block to `set_workspace_resource` docstring would be nice to demonstrate this.